src/pip/_internal/resolution/resolvelib/provider.py

```diff
@@ -0,0 +1,75 @@
+from pip._vendor.resolvelib.providers import AbstractProvider
+
+from pip._internal.utils.typing import MYPY_CHECK_RUNNING
+
+from .requirements import make_requirement
+
+if MYPY_CHECK_RUNNING:
+    from typing import Any, Optional, Sequence, Tuple, Union
+
+    from pip._internal.index.package_finder import PackageFinder
+    from pip._internal.operations.prepare import RequirementPreparer
+    from pip._internal.req.req_install import InstallRequirement
+    from pip._internal.resolution.base import InstallRequirementProvider
+
+    from .base import Requirement, Candidate
+
+
+class PipProvider(AbstractProvider):
+    def __init__(
+        self,
+        finder,    # type: PackageFinder
+        preparer,  # type: RequirementPreparer
+        make_install_req  # type: InstallRequirementProvider
+    ):
+        # type: (...) -> None
+        self._finder = finder
+        self._preparer = preparer
+        self._make_install_req = make_install_req
+
+    def make_requirement(self, ireq):
+        # type: (InstallRequirement) -> Requirement
+        return make_requirement(
+            ireq,
+            self._finder,
+            self._preparer,
+            self._make_install_req
+        )
+
+    def get_install_requirement(self, c):
+        # type: (Candidate) -> InstallRequirement
+        return getattr(c, "_ireq", None)
```

Comment:
Eww...

Reply:
The better way is probably adding `Candidate.get_install_requirement()`. But I'm slightly reluctant to commit to that API before we've got a second subclass of `Candidate` to confirm that it's something all candidates can handle...

I can change it, though - it's probably just as easy to adjust the `Candidate` method later as it is to remove this hack.

I'll have a think about it.

Reply:
Ah I see, this is more about convincing Mypy than actually dealing with a missing `_ireq` attribute… In that case I am on board with this.